interpreter/cling/lib/Interpreter/ClingPragmas.cpp

```diff
@@ -73,10 +73,124 @@ namespace {
       m_Interp.loadFile(FileName, true /*allowSharedLib*/);
     }
   };
+
+  class PHAddIncPath: public PragmaHandler {
+    Interpreter& m_Interp;
+
+  public:
+    PHAddIncPath(Interpreter& interp):
+      PragmaHandler("add_include_path"), m_Interp(interp) {}
+
+    void HandlePragma(Preprocessor &PP,
+                      PragmaIntroducerKind Introducer,
+                      Token &FirstToken) override {
+      // TODO: use Diagnostics!
+
+      struct SkipToEOD_t {
+        Preprocessor& m_PP;
+        SkipToEOD_t(Preprocessor& PP): m_PP(PP) {}
+        ~SkipToEOD_t() { m_PP.DiscardUntilEndOfDirective(); }
+      } SkipToEOD(PP);
+
+      Token Tok;
+      PP.Lex(Tok);
+      if (Tok.isNot(tok::l_paren)) {
+        llvm::errs() << "cling::PHAddIncPath: expect '(' after #pragma cling add_include_path!\n";
+        return;
+      }
+      std::string IncPath;
+      if (!PP.LexStringLiteral(Tok, IncPath, "pragma cling add_include_path",
+                               false /*allowMacroExpansion*/)) {
+        // already diagnosed.
+        return;
+      }
+      if (IncPath.find("$") != std::string::npos)
+        m_Interp.replaceEnvVars(IncPath);
+
+      Preprocessor::CleanupAndRestoreCacheRAII cleanupRAII(PP);
+      clang::Parser& P = m_Interp.getParser();
+      Parser::ParserCurTokRestoreRAII savedCurToken(P);
+      // After we have saved the token reset the current one to something which
+      // is safe (semi colon usually means empty decl)
+      Token& CurTok = const_cast<Token&>(P.getCurToken());
+      CurTok.setKind(tok::semi);
+
+      // We can't PushDeclContext, because we go up and the routine that pops
+      // the DeclContext assumes that we drill down always.
+      // We have to be on the global context. At that point we are in a
+      // wrapper function so the parent context must be the global.
+      TranslationUnitDecl* TU
+        = m_Interp.getCI()->getASTContext().getTranslationUnitDecl();
+      Sema::ContextAndScopeRAII pushedDCAndS(m_Interp.getSema(),
+                                             TU, m_Interp.getSema().TUScope);
+
+      Interpreter::PushTransactionRAII pushedT(&m_Interp);
+      m_Interp.AddIncludePath(IncPath);
```

Comment:
This seems to be mostly identical code with tiny differences. Why don't you move this into a common base class or probably even better into a static utility function, and just call it from these three different places?